rfcs/2020-10-15-3480-file-source-rework.md

```diff
@@ -0,0 +1,425 @@
+
+    // If Vector is shutting down, stop processing
+    maybe_shutdown();
+}
+```
+
+Excepting observability and performance, the primary user-facing concerns in the
+issues above are roughly as follows:
+
+* Read scheduling
+* File identity
+* File starting point
+
+These concerns are a good guide for determining which parts of the
+implementation tend to change together.  Our first goal is to isolate those
+areas of change from one another and construct subcomponents that effectively
+contain complexity. With those seams in place, it should become simpler to
+improve the internals of each.
+
+### Subcomponents
+
+Based on the user-level concerns we collected from the list of issues, let's
+break down what each respective subcomponent should look like.
+
+#### Read scheduling
+
+In the pseudocode above, read scheduling is controlled by `sort`, `should_read`,
+`limit_reached`, `should_not_read_next_file`, and `maybe_backoff`. A component
+would need to answer the following questions:
+
+1. In what order should I read the available files?
+1. Should I finish one file before moving on to the next?
+1. Should I back off reads to this file?
+1. Should I back off reads to all files (i.e. sleep)?
+1. How long should I spend working on a single file?
+
+These questions would then be answered by a combination of configuration, file
+metadata, and gathered statistics.
+
+There are (at least) two approaches we could take to building this component.
+The most obvious would be to implement a struct that contains configuration and
+exposes methods very much like the ones mentioned above. This would maintain the
+current structure of one main read loop with multiple points of control. An
+alternative would be to introduce a trait representing the logic of the read
+loop. This may lead to some duplication, but would allow simpler separation of
+different use cases and require readers of the code understand the interaction
+between fewer subtle points of control.
+
+I would propose that we begin with the simpler method of consolidating logic
+from the existing structure. This involves fewer design decisions and we may
+find with the other simplifications we're planning make that further simplifying
+the read loop is not worthwhile. That being said, once related features (like
+the batch mode discussed later) are complete, we can reevaluate this decision.
+
+#### File identity
+
+Identity is an area where we already have the beginning of modularity with the
+`Fingerprinter` abstraction, but it's incomplete. In our pseudocode, identity is
+involved in both `look_for_files` and `reconcile`. It's not just a way to figure
+out a magic identifier, but also the logic to update our list of watched files
+based on those identifiers. It needs to answer the following:
+
+1. Given a visible path, does it contain a file I've seen before?
+1. If I have seen it before, has it been renamed?
+1. If I have seen it before, am I now seeing it in multiple places?
```

Comment:
I'm not clear that it needs to do this. If a user copies a file that is to be sourced from, and this new file is still captured by the glob, why do we not send it as well?

Reply:
I think this is to catch a rename race, where we may briefly observe the same file in two places.

Reply:
Yes, rotation strategies that copy are one case. This is also something users can run into if fingerprints are colliding for undesirable reasons like common headers, so we should at least log that it's happening so they'll have a hint to update their config.